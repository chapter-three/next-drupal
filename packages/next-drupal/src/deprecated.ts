export * from "./deprecated/get-access-token"
export * from "./deprecated/get-menu"
export * from "./deprecated/get-paths"
export * from "./deprecated/get-resource-collection"
export * from "./deprecated/preview"
export * from "./deprecated/get-resource-type"
export * from "./deprecated/get-resource"
export * from "./deprecated/get-search-index"
export * from "./deprecated/get-view"
export * from "./deprecated/translate-path"
export {
  deserialize,
  buildUrl,
  getJsonApiIndex,
  getJsonApiPathForResourceType,
  syncDrupalPreviewRoutes,
} from "./deprecated/utils"

export type * from "./types/deprecated"
