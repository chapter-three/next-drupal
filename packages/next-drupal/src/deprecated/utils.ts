import { Jsona } from "jsona"
import { stringify } from "qs"
import { getAccessToken } from "./get-access-token"
import type { GetStaticPropsContext } from "next"
import type { AccessToken, Locale } from "../types"

const JSONAPI_PREFIX = process.env.DRUPAL_JSONAPI_PREFIX || "/jsonapi"

const dataFormatter = new Jsona()

export function deserialize(body, options?) {
  if (!body) return null

  return dataFormatter.deserialize(body, options)
}

export async function getJsonApiPathForResourceType(
  type: string,
  locale?: Locale
) {
  const index = await getJsonApiIndex(locale)

  return index?.links[type]?.href
}

export async function getJsonApiIndex(
  locale?: Locale,
  options?: {
    accessToken?: AccessToken
  }
): Promise<{
  links: {
    [type: string]: {
      href: string
    }
  }
}> {
  const url = buildUrl(
    locale ? `/${locale}${JSONAPI_PREFIX}` : `${JSONAPI_PREFIX}`
  )

  // As per https://www.drupal.org/node/2984034 /jsonapi is public.
  // We only call buildHeaders if accessToken or locale is explicitly set.
  // This is for rare cases where /jsonapi might be protected.
  const response = await fetch(url.toString(), {
    headers:
      locale || options
        ? await buildHeaders(options)
        : {
            "Content-Type": "application/json",
          },
  })

  if (!response.ok) {
    throw new Error(response.statusText)
  }

  return await response.json()
}

export function buildUrl(
  path: string,
  params?: string | Record<string, string> | URLSearchParams
): URL {
  const url = new URL(
    path.charAt(0) === "/"
      ? `${process.env.NEXT_PUBLIC_DRUPAL_BASE_URL}${path}`
      : path
  )

  if (params) {
    // Use instead URLSearchParams for nested params.
    url.search = stringify(params)
  }

  return url
}

export async function buildHeaders({
  accessToken,
  headers = {
    "Content-Type": "application/json",
  },
}: {
  accessToken?: AccessToken
  headers?: RequestInit["headers"]
} = {}): Promise<RequestInit["headers"]> {
  // This allows an access_token (preferrably long-lived) to be set directly on the env.
  // This reduces the number of OAuth call to the Drupal server.
  // Intentionally marked as unstable for now.
  if (process.env.UNSTABLE_DRUPAL_ACCESS_TOKEN) {
    headers["Authorization"] =
      `Bearer ${process.env.UNSTABLE_DRUPAL_ACCESS_TOKEN}`

    return headers
  }

  const token = accessToken || (await getAccessToken())
  if (token) {
    headers["Authorization"] = `Bearer ${token.access_token}`
  }

  return headers
}

export function getPathFromContext(
  context: GetStaticPropsContext,
  prefix = ""
) {
  let { slug } = context.params

  slug = Array.isArray(slug)
    ? slug.map((s) => encodeURIComponent(s)).join("/")
    : slug

  // Handle locale.
  if (context.locale && context.locale !== context.defaultLocale) {
    slug = `/${context.locale}/${slug}`
  }

  return !slug
    ? process.env.DRUPAL_FRONT_PAGE
    : prefix
      ? `${prefix}/${slug}`
      : slug
}

export function syncDrupalPreviewRoutes(path) {
  if (window && window.top !== window.self) {
    window.parent.postMessage(
      { type: "NEXT_DRUPAL_ROUTE_SYNC", path },
      process.env.NEXT_PUBLIC_DRUPAL_BASE_URL
    )
  }
}
