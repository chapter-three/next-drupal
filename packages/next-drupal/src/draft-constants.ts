export const DRAFT_DATA_COOKIE_NAME = "next_drupal_draft_data"

// See https://vercel.com/docs/workflow-collaboration/draft-mode
export const DRAFT_MODE_COOKIE_NAME = "__prerender_bypass"
