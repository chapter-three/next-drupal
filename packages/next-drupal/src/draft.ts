import { cookies, draftMode } from "next/headers"
import { redirect } from "next/navigation"
import {
  DRAFT_DATA_COOKIE_NAME,
  DRAFT_MODE_COOKIE_NAME,
} from "./draft-constants"
import type { NextRequest } from "next/server"
import type { NextDrupal } from "./next-drupal"

export async function enableDraftMode(
  request: NextRequest,
  drupal: NextDrupal
): Promise<Response | never> {
  // Validate the draft request.
  const response = await drupal.validateDraftUrl(request.nextUrl.searchParams)

  // If validation fails, don't enable draft mode.
  if (!response.ok) {
    return response
  }

  const searchParams = request.nextUrl.searchParams
  const path = searchParams.get("path")

  // Enable Draft Mode by setting the cookie
  draftMode().enable()

  // Override the default SameSite=lax.
  // See https://github.com/vercel/next.js/issues/49927
  const draftModeCookie = cookies().get(DRAFT_MODE_COOKIE_NAME)
  if (draftModeCookie) {
    cookies().set({
      ...draftModeCookie,
      sameSite: "none",
      secure: true,
    })
  }

  // Send Drupal's data to the draft-mode page.
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { secret, scope, plugin, ...draftData } = Object.fromEntries(
    searchParams.entries()
  )
  cookies().set({
    ...draftModeCookie,
    name: DRAFT_DATA_COOKIE_NAME,
    sameSite: "none",
    secure: true,
    value: JSON.stringify(draftData),
  })

  // Redirect to the path from the fetched post. We can safely redirect to the
  // path since this has been validated on the server.
  redirect(path)
}

export function disableDraftMode() {
  cookies().delete(DRAFT_DATA_COOKIE_NAME)
  draftMode().disable()

  return new Response("Draft mode is disabled")
}

export interface DraftData {
  path?: string
  resourceVersion?: string
}

export function getDraftData() {
  let data: DraftData = {}

  if (draftMode().isEnabled && cookies().has(DRAFT_DATA_COOKIE_NAME)) {
    data = JSON.parse(cookies().get(DRAFT_DATA_COOKIE_NAME)?.value || "{}")
  }

  return data
}
