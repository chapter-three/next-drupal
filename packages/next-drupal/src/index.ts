export * from "./client"
export * from "./jsonapi-errors"

export type * from "./types"

export * from "./deprecated"
