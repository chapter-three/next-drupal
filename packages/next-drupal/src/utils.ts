export { buildHeaders } from "./utils/build-headers"
export { buildUrl } from "./utils/build-url"
export { deserialize } from "./utils/deserialize"
export { getJsonApiIndex } from "./utils/get-json-api-index"
export { getJsonApiPathForResourceType } from "./utils/get-json-api-path-for-resource-type"
export { getPathFromContext } from "./utils/get-path-from-context"
export { syncDrupalPreviewRoutes } from "./utils/sync-drupal-preview-routes"
