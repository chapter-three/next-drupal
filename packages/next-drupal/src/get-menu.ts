import {
  AccessToken,
  DrupalMenuLinkContent,
  FetchAPI,
  JsonApiWithLocaleOptions,
} from "./types"
import { buildHeaders, buildUrl, deserialize } from "./utils"

export async function getMenu(
  name: string,
  options?: {
    deserialize?: boolean
    accessToken?: AccessToken
    fetch?: FetchAPI
  } & JsonApiWithLocaleOptions
): Promise<{
  items: DrupalMenuLinkContent[]
  tree: DrupalMenuLinkContent[]
}> {
  options = {
    deserialize: true,
    fetch,
    ...options,
  }

  const localePrefix =
    options?.locale && options.locale !== options.defaultLocale
      ? `/${options.locale}`
      : ""

  const url = buildUrl(`${localePrefix}/jsonapi/menu_items/${name}`)

  const response = await options.fetch(url.toString(), {
    headers: await buildHeaders(options),
  })

  if (!response.ok) {
    throw new Error(response.statusText)
  }

  const data = await response.json()

  const items = options.deserialize ? deserialize(data) : data

  const { items: tree } = buildMenuTree(items)

  return {
    items,
    tree,
  }
}

function buildMenuTree(
  links: DrupalMenuLinkContent[],
  parent: DrupalMenuLinkContent["id"] = ""
) {
  const children = links.filter((link) => link.parent === parent)

  return children.length
    ? {
        items: children.map((link) => ({
          ...link,
          ...buildMenuTree(links, link.id),
        })),
      }
    : {}
}
