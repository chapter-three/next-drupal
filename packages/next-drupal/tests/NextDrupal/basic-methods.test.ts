import { afterEach, describe, expect, jest, test } from "@jest/globals"
import { NextDrupal } from "../../src"
import { BASE_URL, mockLogger } from "../utils"
import type { DrupalNode, JsonDeserializer } from "../../src"

jest.setTimeout(10000)

afterEach(() => {
  jest.restoreAllMocks()
})

describe("deserialize()", () => {
  test("deserializes JSON:API resource", async () => {
    const drupal = new NextDrupal(BASE_URL)
    const url = await drupal.buildEndpoint({
      resourceType: "node--article",
      path: "/52837ad0-f218-46bd-a106-5710336b7053",
      searchParams: {
        include: "field_tags",
      },
    })

    const response = await drupal.fetch(url)
    expect(response.status).toBe(200)
    const json = await response.json()
    const article = drupal.deserialize(json) as DrupalNode

    expect(article).toMatchSnapshot()
    expect(article.id).toEqual("52837ad0-f218-46bd-a106-5710336b7053")
    expect(article.field_tags).toHaveLength(3)
  })

  test("deserializes JSON:API collection", async () => {
    const drupal = new NextDrupal(BASE_URL)
    const url = await drupal.buildEndpoint({
      resourceType: "node--article",
      searchParams: {
        getQueryObject: () => ({
          "fields[node--article]": "title",
        }),
      },
    })

    const response = await drupal.fetch(url)
    expect(response.status).toBe(200)
    const json = await response.json()
    const articles = drupal.deserialize(json) as DrupalNode[]

    expect(articles).toMatchSnapshot()
  })

  test("allows for custom data serializer", async () => {
    const deserializer: JsonDeserializer = (
      body: { data: { id: string; attributes: { title: string } } },
      options: { pathPrefix: string }
    ) => {
      return {
        id: body.data.id,
        title: `${options.pathPrefix}: ${body.data.attributes.title}`,
      }
    }
    const drupal = new NextDrupal(BASE_URL, {
      deserializer,
    })
    const url = await drupal.buildEndpoint({
      resourceType: "node--article",
      path: "/52837ad0-f218-46bd-a106-5710336b7053",
    })

    const response = await drupal.fetch(url)
    expect(response.status).toBe(200)
    const json = await response.json()
    const article = drupal.deserialize(json, {
      pathPrefix: "TITLE",
    }) as DrupalNode

    expect(article).toMatchSnapshot()
    expect(article.id).toEqual("52837ad0-f218-46bd-a106-5710336b7053")
    expect(article.title).toEqual(`TITLE: ${json.data.attributes.title}`)
  })

  test("returns null if no body", () => {
    const drupal = new NextDrupal(BASE_URL)
    expect(drupal.deserialize("")).toBe(null)
  })
})

describe("logOrThrowError()", () => {
  test("throws the error", () => {
    const drupal = new NextDrupal(BASE_URL)
    expect(() => {
      drupal.logOrThrowError(new Error("Example error"))
    }).toThrow("Example error")
  })

  test("logs the error when throwJsonApiErrors is false", () => {
    const logger = mockLogger()
    const drupal = new NextDrupal(BASE_URL, {
      throwJsonApiErrors: false,
      logger,
    })
    expect(() => {
      drupal.logOrThrowError(new Error("Example error"))
    }).not.toThrow()
    expect(logger.error).toHaveBeenCalledWith(new Error("Example error"))
  })
})
