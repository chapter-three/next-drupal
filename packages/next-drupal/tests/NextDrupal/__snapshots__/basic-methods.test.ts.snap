// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserialize() allows for custom data serializer 1`] = `
{
  "id": "52837ad0-f218-46bd-a106-5710336b7053",
  "title": "TITLE: Give it a go and grow your own herbs",
}
`;

exports[`deserialize() deserializes JSON:API collection 1`] = `
[
  {
    "id": "52837ad0-f218-46bd-a106-5710336b7053",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/52837ad0-f218-46bd-a106-5710336b7053?resourceVersion=id%3A20",
      },
    },
    "title": "Give it a go and grow your own herbs",
    "type": "node--article",
  },
  {
    "id": "3d5a7bca-5b4a-49ee-87d2-cbe313562903",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/3d5a7bca-5b4a-49ee-87d2-cbe313562903?resourceVersion=id%3A22",
      },
    },
    "title": "Dairy-free and delicious milk chocolate",
    "type": "node--article",
  },
  {
    "id": "a1ef61c9-ed15-4b6d-bba2-bf6e5141f961",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/a1ef61c9-ed15-4b6d-bba2-bf6e5141f961?resourceVersion=id%3A24",
      },
    },
    "title": "The real deal for supermarket savvy shopping",
    "type": "node--article",
  },
  {
    "id": "ac0b8b56-db8d-4322-9b1e-224f22083f0d",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/ac0b8b56-db8d-4322-9b1e-224f22083f0d?resourceVersion=id%3A26",
      },
    },
    "title": "The Umami guide to our favorite mushrooms",
    "type": "node--article",
  },
  {
    "id": "22874ecc-0443-441b-a3c9-3aa94d85b800",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/22874ecc-0443-441b-a3c9-3aa94d85b800?resourceVersion=id%3A28",
      },
    },
    "title": "Let's hear it for carrots",
    "type": "node--article",
  },
  {
    "id": "ec51b1de-f51e-4761-b061-aa2624e58b4a",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/ec51b1de-f51e-4761-b061-aa2624e58b4a?resourceVersion=id%3A30",
      },
    },
    "title": "Baking mishaps - our troubleshooting tips",
    "type": "node--article",
  },
  {
    "id": "d5b9ef80-c0aa-45ca-88e7-79c28abe5d50",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/d5b9ef80-c0aa-45ca-88e7-79c28abe5d50?resourceVersion=id%3A32",
      },
    },
    "title": "Skip the spirits with delicious mocktails",
    "type": "node--article",
  },
  {
    "id": "6c7c249f-cd1a-41d5-944b-b3bfce68b325",
    "links": {
      "self": {
        "href": "https://tests.next-drupal.org/en/jsonapi/node/article/6c7c249f-cd1a-41d5-944b-b3bfce68b325?resourceVersion=id%3A34",
      },
    },
    "title": "Give your oatmeal the ultimate makeover",
    "type": "node--article",
  },
]
`;

exports[`deserialize() deserializes JSON:API resource 1`] = `
{
  "body": {
    "format": "basic_html",
    "processed": "<p>There's nothing like having your own supply of fresh herbs, readily available and close at hand to use while cooking. Whether you have a large garden or a small kitchen window sill, there's always enough room for something home grown.</p>
<h2>Outdoors</h2>
<h3>Mint</h3>
<p>Mint is a great plant to grow as it's hardy and can grow in almost any soil. Mint can grow wild, so keep it contained in a pot or it might spread and take over your whole garden.</p>
<h3>Sage</h3>
<p>Like mint, sage is another prolific growing plant and will take over your garden if you let it. Highly aromatic, the sage plant can be planted in a pot or flower bed in well drained soil. The best way to store the herb is to sun dry the leaves and store in a cool, dark cupboard in a sealed container.</p>
<h3>Rosemary</h3>
<p>Rosemary plants grow into lovely shrubs. Easily grown from cuttings, rosemary plants do not like freezing temperatures so keep pots or planted bushes near the home to shelter them from the cold. It grows well in pots as it likes dry soil, but can survive well in the ground too. If pruning rosemary to encourage it into a better shape, save the branches and hang them upside down to preserve the flavor and use in food.</p>
<h2>Indoors</h2>
<h3>Basil</h3>
<p>Perfect in sunny spot on a kitchen window sill. Basil is an annual plant, so will die off in the autumn, so it's a good idea to harvest it in the summer if you have an abundance and dry it. Picked basil stays fresh longer if it is placed in water (like fresh flowers). A great way to store basil is to make it into pesto!</p>
<h3>Chives</h3>
<p>A versatile herb, chives can grow well indoors. Ensure the plant is watered well, and gets plenty of light. Remember to regularly trim the chives. This prevents the flowers from developing and encourages new growth.</p>
<h3>Coriander (Cilantro)</h3>
<p>Coriander can grow indoors, but unlike the other herbs, it doesn't like full sun. If you have a south facing kitchen window, this isn't the place for it. Although not as thirsty as basil, coriander doesn't like dry soil so don't forget to water it! Cut coriander is best stored in the fridge.</p>
",
    "summary": null,
    "value": "<p>There's nothing like having your own supply of fresh herbs, readily available and close at hand to use while cooking. Whether you have a large garden or a small kitchen window sill, there's always enough room for something home grown.</p>
<h2>Outdoors</h2>
<h3>Mint</h3>
<p>Mint is a great plant to grow as it's hardy and can grow in almost any soil. Mint can grow wild, so keep it contained in a pot or it might spread and take over your whole garden.</p>
<h3>Sage</h3>
<p>Like mint, sage is another prolific growing plant and will take over your garden if you let it. Highly aromatic, the sage plant can be planted in a pot or flower bed in well drained soil. The best way to store the herb is to sun dry the leaves and store in a cool, dark cupboard in a sealed container.</p>
<h3>Rosemary</h3>
<p>Rosemary plants grow into lovely shrubs. Easily grown from cuttings, rosemary plants do not like freezing temperatures so keep pots or planted bushes near the home to shelter them from the cold. It grows well in pots as it likes dry soil, but can survive well in the ground too. If pruning rosemary to encourage it into a better shape, save the branches and hang them upside down to preserve the flavor and use in food.</p>
<h2>Indoors</h2>
<h3>Basil</h3>
<p>Perfect in sunny spot on a kitchen window sill. Basil is an annual plant, so will die off in the autumn, so it's a good idea to harvest it in the summer if you have an abundance and dry it. Picked basil stays fresh longer if it is placed in water (like fresh flowers). A great way to store basil is to make it into pesto!</p>
<h3>Chives</h3>
<p>A versatile herb, chives can grow well indoors. Ensure the plant is watered well, and gets plenty of light. Remember to regularly trim the chives. This prevents the flowers from developing and encourages new growth.</p>
<h3>Coriander (Cilantro)</h3>
<p>Coriander can grow indoors, but unlike the other herbs, it doesn't like full sun. If you have a south facing kitchen window, this isn't the place for it. Although not as thirsty as basil, coriander doesn't like dry soil so don't forget to water it! Cut coriander is best stored in the fridge.</p>
",
  },
  "changed": "2022-03-21T10:52:42+00:00",
  "content_translation_outdated": false,
  "content_translation_source": "und",
  "created": "2022-03-21T10:52:42+00:00",
  "default_langcode": true,
  "drupal_internal__nid": 10,
  "drupal_internal__vid": 20,
  "field_media_image": {
    "id": "e5091a16-134e-400d-8393-cfe4eccbcaa2",
    "resourceIdObjMeta": {
      "drupal_internal__target_id": 10,
    },
    "type": "media--image",
  },
  "field_tags": [
    {
      "changed": "2022-03-21T10:52:42+00:00",
      "content_translation_created": "2022-03-21T10:52:42+00:00",
      "content_translation_outdated": false,
      "content_translation_source": "und",
      "content_translation_uid": {
        "id": "256a133b-0bd7-4426-a823-b8ce81e0d778",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": 0,
        },
        "type": "user--user",
      },
      "default_langcode": true,
      "description": null,
      "drupal_internal__revision_id": 14,
      "drupal_internal__tid": 14,
      "id": "dcd81647-71b7-48cb-b555-e20322bcb7a7",
      "langcode": "en",
      "links": {
        "self": {
          "href": "https://tests.next-drupal.org/en/jsonapi/taxonomy_term/tags/dcd81647-71b7-48cb-b555-e20322bcb7a7?resourceVersion=id%3A14",
        },
      },
      "name": "Grow your own",
      "parent": [
        {
          "id": "virtual",
          "resourceIdObjMeta": {
            "links": {
              "help": {
                "href": "https://www.drupal.org/docs/8/modules/json-api/core-concepts#virtual",
                "meta": {
                  "about": "Usage and meaning of the 'virtual' resource identifier.",
                },
              },
            },
          },
          "type": "taxonomy_term--tags",
        },
      ],
      "path": {
        "alias": "/tags/grow-your-own",
        "langcode": "en",
        "pid": 27,
      },
      "relationshipNames": [
        "vid",
        "revision_user",
        "parent",
        "content_translation_uid",
      ],
      "resourceIdObjMeta": {
        "drupal_internal__target_id": 14,
      },
      "revision_created": "2022-03-21T10:52:42+00:00",
      "revision_log_message": null,
      "revision_translation_affected": true,
      "revision_user": null,
      "status": true,
      "type": "taxonomy_term--tags",
      "vid": {
        "id": "b0442737-e5ff-4c7c-8c2b-c977886d6e73",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": "tags",
        },
        "type": "taxonomy_vocabulary--taxonomy_vocabulary",
      },
      "weight": 0,
    },
    {
      "changed": "2022-03-21T10:52:42+00:00",
      "content_translation_created": "2022-03-21T10:52:42+00:00",
      "content_translation_outdated": false,
      "content_translation_source": "und",
      "content_translation_uid": {
        "id": "256a133b-0bd7-4426-a823-b8ce81e0d778",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": 0,
        },
        "type": "user--user",
      },
      "default_langcode": true,
      "description": null,
      "drupal_internal__revision_id": 23,
      "drupal_internal__tid": 23,
      "id": "60d20a4c-9d42-4b25-b717-3af3cba6abe8",
      "langcode": "en",
      "links": {
        "self": {
          "href": "https://tests.next-drupal.org/en/jsonapi/taxonomy_term/tags/60d20a4c-9d42-4b25-b717-3af3cba6abe8?resourceVersion=id%3A23",
        },
      },
      "name": "Seasonal",
      "parent": [
        {
          "id": "virtual",
          "resourceIdObjMeta": {
            "links": {
              "help": {
                "href": "https://www.drupal.org/docs/8/modules/json-api/core-concepts#virtual",
                "meta": {
                  "about": "Usage and meaning of the 'virtual' resource identifier.",
                },
              },
            },
          },
          "type": "taxonomy_term--tags",
        },
      ],
      "path": {
        "alias": "/tags/seasonal",
        "langcode": "en",
        "pid": 45,
      },
      "relationshipNames": [
        "vid",
        "revision_user",
        "parent",
        "content_translation_uid",
      ],
      "resourceIdObjMeta": {
        "drupal_internal__target_id": 23,
      },
      "revision_created": "2022-03-21T10:52:42+00:00",
      "revision_log_message": null,
      "revision_translation_affected": true,
      "revision_user": null,
      "status": true,
      "type": "taxonomy_term--tags",
      "vid": {
        "id": "b0442737-e5ff-4c7c-8c2b-c977886d6e73",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": "tags",
        },
        "type": "taxonomy_vocabulary--taxonomy_vocabulary",
      },
      "weight": 0,
    },
    {
      "changed": "2022-03-21T10:52:42+00:00",
      "content_translation_created": "2022-03-21T10:52:42+00:00",
      "content_translation_outdated": false,
      "content_translation_source": "und",
      "content_translation_uid": {
        "id": "256a133b-0bd7-4426-a823-b8ce81e0d778",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": 0,
        },
        "type": "user--user",
      },
      "default_langcode": true,
      "description": null,
      "drupal_internal__revision_id": 16,
      "drupal_internal__tid": 16,
      "id": "57a1d9f6-23a6-4215-a8a9-582202cd938d",
      "langcode": "en",
      "links": {
        "self": {
          "href": "https://tests.next-drupal.org/en/jsonapi/taxonomy_term/tags/57a1d9f6-23a6-4215-a8a9-582202cd938d?resourceVersion=id%3A16",
        },
      },
      "name": "Herbs",
      "parent": [
        {
          "id": "virtual",
          "resourceIdObjMeta": {
            "links": {
              "help": {
                "href": "https://www.drupal.org/docs/8/modules/json-api/core-concepts#virtual",
                "meta": {
                  "about": "Usage and meaning of the 'virtual' resource identifier.",
                },
              },
            },
          },
          "type": "taxonomy_term--tags",
        },
      ],
      "path": {
        "alias": "/tags/herbs",
        "langcode": "en",
        "pid": 31,
      },
      "relationshipNames": [
        "vid",
        "revision_user",
        "parent",
        "content_translation_uid",
      ],
      "resourceIdObjMeta": {
        "drupal_internal__target_id": 16,
      },
      "revision_created": "2022-03-21T10:52:42+00:00",
      "revision_log_message": null,
      "revision_translation_affected": true,
      "revision_user": null,
      "status": true,
      "type": "taxonomy_term--tags",
      "vid": {
        "id": "b0442737-e5ff-4c7c-8c2b-c977886d6e73",
        "resourceIdObjMeta": {
          "drupal_internal__target_id": "tags",
        },
        "type": "taxonomy_vocabulary--taxonomy_vocabulary",
      },
      "weight": 0,
    },
  ],
  "id": "52837ad0-f218-46bd-a106-5710336b7053",
  "langcode": "en",
  "links": {
    "self": {
      "href": "https://tests.next-drupal.org/en/jsonapi/node/article/52837ad0-f218-46bd-a106-5710336b7053?resourceVersion=id%3A20",
    },
  },
  "moderation_state": "published",
  "node_type": {
    "id": "a145b65a-e660-4f5d-ac0d-bd2ff9e3f0b0",
    "resourceIdObjMeta": {
      "drupal_internal__target_id": "article",
    },
    "type": "node_type--node_type",
  },
  "path": {
    "alias": "/articles/give-it-a-go-and-grow-your-own-herbs",
    "langcode": "en",
    "pid": 85,
  },
  "promote": true,
  "relationshipNames": [
    "node_type",
    "revision_uid",
    "uid",
    "field_media_image",
    "field_tags",
  ],
  "revision_log": null,
  "revision_timestamp": "2022-03-21T10:52:42+00:00",
  "revision_translation_affected": null,
  "revision_uid": {
    "id": "dd9c916d-4d66-4bff-a851-eeba0cf7673a",
    "resourceIdObjMeta": {
      "drupal_internal__target_id": 5,
    },
    "type": "user--user",
  },
  "status": true,
  "sticky": false,
  "title": "Give it a go and grow your own herbs",
  "type": "node--article",
  "uid": {
    "id": "dd9c916d-4d66-4bff-a851-eeba0cf7673a",
    "resourceIdObjMeta": {
      "drupal_internal__target_id": 5,
    },
    "type": "user--user",
  },
}
`;
